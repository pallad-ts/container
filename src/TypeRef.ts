import * as isPred from "predicates";
import { DefinitionPredicate } from "./types";

const reservedConstructors = new Set<Function>(
	["Function", "Object", "Promise"].map(x => (global as any)[x]).filter(x => x)
);

export class TypeRef {
	constructor(private target: Function) {
		if (!TypeRef.isAllowedTarget(target)) {
			throw new Error(`Target ${target} is not allowed`);
		}
		Object.freeze(this);
	}

	matches(type: TypeRef): boolean {
		//tslint:disable-next-line: strict-comparisons
		return type.target === this.target || type.target.prototype instanceof this.target;
	}

	toString() {
		return `instance of class "${this.target.name}"`;
	}

	get predicate(): DefinitionPredicate {
		return x => x.type !== undefined && this.matches(x.type);
	}

	static isAllowedTarget(target: Function) {
		return !reservedConstructors.has(target);
	}

	static is(value: any): value is TypeRef {
		return value instanceof TypeRef;
	}

	static createFromType(type: Function): TypeRef | undefined {
		if (TypeRef.isAllowedTarget(type)) {
			return new TypeRef(type);
		}
	}

	static createFromValue(value: any) {
		// eslint-disable-next-line no-null/no-null
		if (isPred.object(value) && value !== null) {
			const proto = Object.getPrototypeOf(value);
			if (TypeRef.isAllowedTarget(proto.constructor)) {
				return new TypeRef(proto.constructor);
			}
		}
	}

	static predicateForType(type: Function): DefinitionPredicate | undefined {
		const ref = TypeRef.createFromType(type);
		if (ref) {
			return ref.predicate;
		}
	}
}
