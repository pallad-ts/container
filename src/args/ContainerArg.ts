import {Container} from '../Container';
import {Definition} from '../Definition';

export abstract class ContainerArg<T = any> {
    /**
     * Returns argument value
     */
    abstract getArgument(container: Container): Promise<T>;

    /**
     * Returns a definition or list of definitions of services that giver argument requires
     */
    abstract getDependentServices(container: Container): Definition | Definition[];

    static is(value: any): value is ContainerArg {
        return value instanceof ContainerArg;
    }
}