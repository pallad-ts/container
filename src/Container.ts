import {
    AnnotationPredicate, ServiceName, DefinitionPredicate, Middleware,
    ServiceFactory, onMiddlewareAttach
} from './types';
import {Definition} from './Definition';
import * as errors from './errors';
import {assertNoCircularDependencies} from './assertNoCircularDependencies';
import {ContainerArg} from './args/ContainerArg';
import * as is from 'predicates';
import {randomName} from './randomName';
import {debugFn} from './debugFn';

const debugCreation = debugFn('creation');
const debugDefinition = debugFn('definition');
const debugSlowCreation = debugFn('slow-creation');

export class Container {
    private definitions: Map<string | Symbol, Definition> = new Map();
    private services: Map<Definition, Promise<any>> = new Map();
    private middlewares: Middleware[] = [];

    public readonly parent?: Container;

    /**
     * Time needed to trigger debug slow creation log
     */
    public slowLogThreshold: number = 10000;

    constructor(parent?: Container) {
        Object.defineProperty(this, 'parent', {
            value: parent,
            writable: false
        });
    }

    /**
     * Registers given service definition
     */
    registerDefinition(definition: Definition): this {
        if (this.definitions.has(definition.name)) {
            throw errors.ALREADY_DEFINED(`Service "${definition.name.toString()}" already defined`);
        }
        debugDefinition(`Service ${definition.name.toString()} defined`);
        definition.setOwner(this);
        this.definitions.set(definition.name, definition);
        return this;
    }

    /**
     * Creates and registers service definition
     *
     * Returns created definition for further configuration
     */
    definition(name?: ServiceName) {
        const definition = new Definition(name);
        this.registerDefinition(definition);
        return definition;
    }

    /**
     * Creates and registers service definition with given name, function as constructor
     */
    definitionWithConstructor(name: ServiceName, clazz: { new(...args: any[]): any }): Definition;
    definitionWithConstructor(clazz: { new(...args: any[]): any }): Definition;
    definitionWithConstructor(nameOrClazz: ServiceName | { new(...args: any[]): any }, clazz?: { new(...args: any[]): any }) {
        const finalClazz = is.func(nameOrClazz) ? nameOrClazz : clazz!;
        return this.definition(ServiceName.is(nameOrClazz) ? nameOrClazz : randomName(finalClazz.name))
            .useConstructor(finalClazz);
    }

    /**
     * Creates and registers service definition with given name, function as factory
     */
    definitionWithFactory(name: ServiceName, factory: ServiceFactory, type?: Function): Definition;
    definitionWithFactory(factory: ServiceFactory, type?: Function): Definition;
    definitionWithFactory(nameOrFactory: ServiceName | ServiceFactory, factoryOrType: ServiceFactory | Function, type?: Function): Definition {
        const name = ServiceName.is(nameOrFactory) ? nameOrFactory : undefined;
        const factory = is.func(nameOrFactory) ? nameOrFactory as ServiceFactory : factoryOrType as ServiceFactory;
        const finalType = is.func(nameOrFactory) ? factoryOrType : type;

        const def = this.definition(name)
            .useFactory(factory);
        if (finalType) {
            def.markType(finalType);
        }
        return def;
    }

    /**
     * Creates and registers service definition with given name and value as a service
     */
    definitionWithValue(name: ServiceName, value: any): Definition;
    definitionWithValue(value: any): Definition;
    definitionWithValue(nameOrValue: ServiceName | any, value?: any): Definition {
        const isNameProvided = value !== undefined && ServiceName.is(nameOrValue);
        const finalValue = isNameProvided ? value : nameOrValue;
        const name = isNameProvided ? nameOrValue : randomName(Object.getPrototypeOf(finalValue).constructor.name);
        return this.definition(name)
            .useValue(finalValue);
    }

    /**
     * Returns definition by given name
     */
    findByName(name: ServiceName): Definition | undefined {
        if (this.definitions.has(name)) {
            return this.definitions.get(name);
        }

        if (this.parent) {
            return this.parent.findByName(name);
        }
    }

    /**
     * Returns definitions that satisfy given predicate
     */
    findByPredicate(predicate: DefinitionPredicate): Definition[] {
        return Array.from(this.definitions.values())
            .filter(predicate)
            .concat(this.parent ? this.parent.findByPredicate(predicate) : []);
    }


    /**
     * Returns all definitions that contain annotation that satisfied given predicate
     */
    findByAnnotation(predicate: AnnotationPredicate): Definition[];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: false): Definition[];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: true): Array<[Definition, any]>;
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: boolean = false): Definition[] | Array<[Definition, any]> {
        let annotations: any[] = [];
        const definitions = this.findByPredicate(s => {
            const annotation = s.annotations.find(predicate);

            if (annotation) {
                withAnnotation && annotations.push(annotation);
                return true;
            }
            return false;
        });

        if (withAnnotation) {
            return definitions.map((d): [Definition, any] => {
                return [d, annotations.shift()];
            });
        }
        return definitions;
    }

    /**
     * Registers given middleware
     */
    addMiddleware(...middlewares: Middleware[]): Container {
        for (const middleware of middlewares) {
            if (middleware[onMiddlewareAttach]) {
                middleware[onMiddlewareAttach]!(this);
            }
        }
        this.middlewares.push(...middlewares);
        return this;
    }

    getMiddlewares(): Middleware[] {
        return this.middlewares;
    }

    /**
     * Returns service for given name or definition
     */
    get<T = any>(nameOrDefinition: ServiceName | Definition): Promise<T> {
        let definition: Definition | undefined;

        if (ServiceName.is(nameOrDefinition)) {
            definition = this.findByName(nameOrDefinition);
            if (definition === undefined) {
                return Promise.reject(errors.SERVICE_NOT_FOUND(`Service "${nameOrDefinition.toString()}" does not exist`));
            }
        } else {
            definition = nameOrDefinition;
        }

        if (this.services.has(definition)) {
            return this.services.get(definition)!;
        }

        if (!this.definitions.has(definition.name) && this.parent) {
            return this.parent.get(definition);
        }

        const promise = this.create(definition);
        this.services.set(definition, promise);
        return promise;
    }

    private async create(definition: Definition) {
        if (!definition.factory) {
            return Promise.reject(
                errors.INCOMPLETE_DEFINITION(
                    `Missing factory for service definition "${definition.name.toString()}". Define it as constructor, factory or value`
                )
            );
        }

        assertNoCircularDependencies(this, definition);

        // valid definition, time to lock it
        definition.lock();

        let timeout: any;
        if (this.slowLogThreshold > 0) {
            timeout = setTimeout(() => {
                debugSlowCreation(`Service ${definition.name.toString()} takes a long time to create. Over ${this.slowLogThreshold} ms`);
            }, this.slowLogThreshold);
        }

        const debugMsg = `Creating service ${definition.name.toString()}`;
        debugCreation(`${debugMsg} - started`);
        let currentMiddleware = 0;
        const middlewares = this.getMiddlewares();
        const next = (definition: Definition) => {
            const middleware = middlewares[currentMiddleware];
            currentMiddleware++;
            if (middleware) {
                return middleware.call(this, definition, next);
            }

            return Promise.all(
                definition.args.map(a => ContainerArg.is(a) ? a.getArgument(this) : a)
            )
                .then((args: any[]) => definition.factory.apply(this, args));
        };

        try {
            const result = await next.call(this, definition);
            timeout && clearTimeout(timeout);
            debugCreation(`${debugMsg} - finished`);
            return result;
        } catch (e) {
            timeout && clearTimeout(timeout);
            throw e;
        }
    }

    /**
     * Returns all services that definition satisfies predicate
     */
    getByPredicate<T = any>(predicate: DefinitionPredicate): Promise<T[]> {
        return Promise.all(
            this.findByPredicate(predicate)
                .map(d => this.get(d))
        );
    }

    /**
     * Returns all services that definition contains annotation that satisfies given predicate
     */
    getByAnnotation<T = any>(predicate: AnnotationPredicate): Promise<T[]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: false): Promise<T[]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: true): Promise<Array<[T, any]>>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: boolean = false): Promise<T[] | Array<[T, any]>> {
        return Promise.all(
            this.findByAnnotation(predicate, true)
                .map(async ([definition, annotation]) => {
                    if (withAnnotation) {
                        return [await this.get(definition), annotation];
                    }
                    return this.get(definition);
                })
        );
    }

    alias(definition: Definition, options?: {
        name?: string,
        withAnnotations?: boolean | ((ann: any) => boolean)
    }) {
        const def = this.definitionWithFactory(options?.name ? options.name : definition.name, () => {
            if (!definition.owner) {
                throw errors.DEFINITION_WITHOUT_CONTAINER(
                    `Cannot create service "${def.name.toString()}" due to lack of assigned container`
                );
            }
            return definition.owner.get(definition);
        });

        const annotations = (() => {
            const withAnnotations = options?.withAnnotations ?? false;
            if (is.func(withAnnotations)) {
                return definition.annotations.filter(withAnnotations);
            }
            return withAnnotations ? definition.annotations : [];
        })();

        if (annotations.length > 0) {
            def.annotate(...annotations);
        }
        if (definition.type) {
            def.markType(definition.type);
        }
        return def;
    }
}