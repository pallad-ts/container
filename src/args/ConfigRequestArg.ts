import * as is from 'predicates';
import {ContainerArg} from "./ContainerArg";
import {Container} from "../Container";
import {getConfigProviderForContainer} from "../middlewares/config";

const assertNotBlank = is.assert(is.notBlank, 'Config "path" cannot be blank');

export class ConfigRequestArg<T> extends ContainerArg<T> {
    public readonly hasDefaultValue: boolean;

    constructor(public readonly path: string, public readonly defaultValue?: any) {
        super();
        assertNotBlank(this.path);
        this.hasDefaultValue = arguments.length > 1;
        Object.freeze(this);
    }

    static create<T = any>(path: string, defaultValue?: T): ConfigRequestArg<T>;
    static create<T = any>(...args: [string, undefined | T]): ConfigRequestArg<T> {
        return new ConfigRequestArg<T>(...args);
    }

    async getArgument(container: Container): Promise<T> {
        return getConfigProviderForContainer(container)(this);
    }

    getDependentServices(container: Container) {
        return [];
    }
}