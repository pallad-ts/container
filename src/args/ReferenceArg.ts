import { AnnotationPredicate, ServiceName, DefinitionPredicate } from "../types";
import { Container } from "../Container";
import { Definition } from "../Definition";
import * as errors from "../errors";
import { Lookup } from "../Lookup";
import { ContainerArg } from "./ContainerArg";
import { TypeRef } from "../TypeRef";
import { ERRORS } from "../errors";

function toTypeRef(type: TypeRef | Function) {
	return TypeRef.is(type) ? type : new TypeRef(type);
}

export class ReferenceArg extends ContainerArg {
	static one = {
		name(name: ServiceName) {
			return new ReferenceArg("one", new Lookup.ByServiceName(name));
		},
		predicate(predicate: DefinitionPredicate) {
			return new ReferenceArg("one", new Lookup.ByPredicate(predicate));
		},
		annotation(predicate: AnnotationPredicate) {
			return new ReferenceArg("one", new Lookup.ByAnnotation(predicate));
		},
		type(type: TypeRef | Function) {
			return new ReferenceArg("one", new Lookup.ByType(toTypeRef(type)));
		},
	};

	static multi = {
		predicate(predicate: DefinitionPredicate) {
			return new ReferenceArg("multi", new Lookup.ByPredicate(predicate));
		},
		annotation(predicate: AnnotationPredicate) {
			return new ReferenceArg("multi", new Lookup.ByAnnotation(predicate));
		},
		type(type: TypeRef | Function) {
			return new ReferenceArg("multi", new Lookup.ByType(toTypeRef(type)));
		},
	};

	constructor(
		private readonly type: "one" | "multi",
		private readonly lookup: Lookup
	) {
		super();
		Object.freeze(this);
	}

	getArgument(container: Container): Promise<any> {
		const definitions = this.findDefinitions(container);
		if (Array.isArray(definitions)) {
			return Promise.all(this.findMulti(container).map(d => container.get(d)));
		}
		return container.get(definitions);
	}

	getDependentServices(container: Container) {
		return this.findDefinitions(container);
	}

	private findDefinitions(container: Container): Definition | Definition[] {
		if (this.type === "one") {
			return this.findOne(container);
		}

		return this.findMulti(container);
	}

	private findOne(container: Container): Definition {
		const definitions = this.lookup.find(container);
		if (Array.isArray(definitions)) {
			if (definitions.length === 0) {
				throw ERRORS.NO_MATCHING_SERVICE.create(this.lookup);
			}

			if (definitions.length > 1) {
				const servicesNames = definitions.map(s => s.name.toString());
				throw ERRORS.AMBIGUOUS_SERVICE.create(servicesNames, this.lookup);
			}
			return definitions[0];
		}
		return definitions;
	}

	private findMulti(container: Container): Definition[] {
		return this.lookup.find(container) as Definition[];
	}
}
