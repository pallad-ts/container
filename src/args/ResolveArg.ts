import {ContainerArg} from "./ContainerArg";
import {Container} from "../Container";
import {Definition} from "../Definition";
import * as is from 'predicates';

const isOmitableType = is.any(is.primitive, is.func);
/**
 * Resolves all ContainerArg in provided value
 */
export class ResolveArg<T extends object> extends ContainerArg<ResolveArg.Resolved<T>> {
    constructor(private readonly value: T | (() => T)) {
        super();
        Object.freeze(this);
    }

    getArgument(container: Container) {
        return this.transformObject(this.getValue(), container);
    }

    private getValue(): T {
        return is.func(this.value) ? this.value() : this.value;
    }

    private async transformObject<T>(result: T, container: Container): Promise<ResolveArg.Resolved<T>> {
        if (isOmitableType(result)) {
            return result as ResolveArg.Resolved<T>;
        }

        if (ContainerArg.is(result)) {
            return result.getArgument(container);
        }

        if (is.arr(result)) {
            return Promise.all(
                result.map(x => isOmitableType(x) ? x : this.transformObject(x, container))
            ) as ResolveArg.Resolved<T>;
        }

        if (is.plainObject(result)) {
            const newObject: any = {};
            for (const [key, value] of Object.entries(result)) {
                newObject[key] = isOmitableType(value) ? value : await this.transformObject(value, container);
            }
            return newObject;
        }

        return result;
    }

    private reducer = (acc: ContainerArg[], current: any) => {
        const args = this.collectContainerArgsFromResult(current);
        if (args) {
            return acc.concat(args);
        }
        return acc;
    };

    private collectContainerArgsFromResult(result: T): ContainerArg[] | undefined {
        if (ContainerArg.is(result)) {
            return [result];
        }

        if (is.primitive(result)) {
            return;
        }

        if (is.array(result)) {
            return result.reduce(this.reducer, []);
        }

        if (is.plainObject(result)) {
            return Object.values(result)
                .reduce(this.reducer, []);
        }
    }

    getDependentServices(container: Container): Definition | Definition[] {
        const value = this.getValue();
        const args = this.collectContainerArgsFromResult(value);
        if (args) {
            return args.reduce<Definition[]>((acc, current) => {
                const deps = current.getDependentServices(container);
                return acc.concat(
                    Array.isArray(deps) ? deps : [deps]
                );
            }, []);
        }
        return [];
    }
}

export namespace ResolveArg {
    export type Resolved<T> = T extends ContainerArg<infer U> ? U :
        (
            T extends Array<infer B> ? Array<Resolved<B>> : (
                T extends object ? {
                    [P in keyof T]: Resolved<T[P]>
                } : T
                )
            );
}