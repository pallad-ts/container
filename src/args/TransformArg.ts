import { ContainerArg } from "./ContainerArg";
import { Container } from "../Container";

/**
 * Applies transformer function on result of provided argument
 */
export class TransformArg<T, TResult> extends ContainerArg<TResult> {
	constructor(
		private readonly transformer: (value: T) => TResult,
		private readonly arg: ContainerArg<T>
	) {
		super();
		Object.freeze(this);
	}

	async getArgument(container: Container): Promise<TResult> {
		return this.transformer(await this.arg.getArgument(container));
	}

	getDependentServices(container: Container) {
		return this.arg.getDependentServices(container);
	}
}
