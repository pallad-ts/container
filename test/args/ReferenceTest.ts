import {Container} from "@src/Container";
import {ReferenceArg} from "@src/args/ReferenceArg";
import {TypeRef} from "@src/TypeRef";

describe('Reference', () => {
    let container: Container;

    const serviceA = {service: 'A'};
    const serviceB = {service: 'B'};
    const serviceD = {service: 'D'};

    const ANNOTATION_NAME = 'annotationName';
    const ANNOTATION = {name: 'ExtraAnnotation'};
    const AMBIGUOUS_ANNOTATION = {name: 'ambiguous'};

    class A {}

    class B {}

    class D extends B {}

    beforeEach(() => {
        container = new Container();

        container.definition('A')
            .useValue(serviceA)
            .markType(A)
            .annotate({name: ANNOTATION_NAME})
            .annotate(AMBIGUOUS_ANNOTATION);

        container.definition('B')
            .useValue(serviceB)
            .markType(B)
            .annotate(ANNOTATION)
            .annotate(AMBIGUOUS_ANNOTATION);

        container.definition('D')
            .useValue(serviceD)
            .markType(D);
    });

    describe('one', () => {
        it('by name', async () => {
            const ref = ReferenceArg.one.name('A');
            expect(ref.getDependentServices(container))
                .toEqual(container.findByName('A'));

            expect(await ref.getArgument(container))
                .toEqual(serviceA);
        });

        it('by name - not found', () => {
            const ref = ReferenceArg.one.name('C');
            expect(() => {
                ref.getDependentServices(container);
            })
                .toThrowError(/No matching service/);
        });

        it('by predicate', async () => {
            const ref = ReferenceArg.one.predicate(d => d.name === 'B');

            expect(ref.getDependentServices(container))
                .toEqual(container.findByName('B'));

            expect(await ref.getArgument(container))
                .toEqual(serviceB);
        });

        it('by predicate - not found', () => {
            const ref = ReferenceArg.one.predicate(() => false);

            expect(() => {
                ref.getDependentServices(container);
            })
                .toThrowError(/No matching service/);
        });

        it('by predicate - ambiguous error', () => {
            const ref = ReferenceArg.one.predicate(() => true);

            expect(() => {
                ref.getDependentServices(container);
            })
                .toThrowError(/Multiple services found \(A, B, D\)/);
        });

        it('by annotation', async () => {
            const ref = ReferenceArg.one.annotation(a => a.name === ANNOTATION_NAME);

            expect(ref.getDependentServices(container))
                .toEqual(container.findByName('A'));

            expect(await ref.getArgument(container))
                .toEqual(serviceA);
        });

        it('by annotation - not found', () => {
            const ref = ReferenceArg.one.annotation(() => false);

            expect(() => {
                ref.getDependentServices(container);
            })
                .toThrowError(/No matching service/);
        });

        it('by annotation- ambiguous error', () => {
            const ref = ReferenceArg.one.annotation(a => a.name === AMBIGUOUS_ANNOTATION.name);

            expect(() => {
                ref.getDependentServices(container);
            })
                .toThrowError(/Multiple services found \(A, B\)/);
        });

        it('by type', () => {
            const ref = ReferenceArg.one.type(TypeRef.createFromType(A)!)

            expect(ref.getDependentServices(container))
                .toEqual(container.findByName('A'));
        });

        it('by type - ambiguous error', () => {
            const ref = ReferenceArg.one.type(TypeRef.createFromType(B)!)
            expect(() => {
                ref.getDependentServices(container)
            })
                .toThrowErrorMatchingSnapshot();
        })
    });

    describe('multi', () => {
        it('by annotation', async () => {
            const ref = ReferenceArg.multi.annotation(a => a.name === AMBIGUOUS_ANNOTATION.name);

            expect(ref.getDependentServices(container))
                .toEqual([
                    container.findByName('A'),
                    container.findByName('B')
                ])

            expect(await ref.getArgument(container))
                .toEqual([serviceA, serviceB]);
        });

        it('by annotation - nothing found', async () => {
            const ref = ReferenceArg.multi.annotation(() => false);

            expect(ref.getDependentServices(container))
                .toHaveLength(0);

            expect(await ref.getArgument(container))
                .toHaveLength(0);
        });

        it('by predicate', async () => {
            const ref = ReferenceArg.multi.predicate(d => d.name === 'A');

            expect(ref.getDependentServices(container))
                .toEqual([
                    container.findByName('A')
                ])

            expect(await ref.getArgument(container))
                .toEqual([serviceA]);
        });

        it('by predicate - nothing found', async () => {
            const ref = ReferenceArg.multi.predicate(() => false);

            expect(ref.getDependentServices(container))
                .toHaveLength(0);
            expect(await ref.getArgument(container))
                .toHaveLength(0);
        });

        it('by type', () => {
            const ref = ReferenceArg.multi.type(TypeRef.createFromType(B)!)

            expect(ref.getDependentServices(container))
                .toEqual([
                    container.findByName('B'),
                    container.findByName('D')
                ]);
        });
    });
});