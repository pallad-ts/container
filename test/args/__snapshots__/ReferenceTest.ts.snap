// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference one by type - ambiguous error 1`] = `"Multiple services found (B, D) with following lookup: by type: instance of class "B""`;
